platform/darwin/src/MLNCustomDrawableStyleLayer.mm

```diff
@@ -0,0 +1,279 @@
+#import "MLNCustomDrawableStyleLayer.h"
+#import "MLNCustomDrawableStyleLayer_Private.h"
+
+#import "MLNMapView_Private.h"
+#import "MLNStyle_Private.h"
+#import "MLNStyleLayer_Private.h"
+#import "MLNGeometry_Private.h"
+
+#include <mbgl/style/layers/custom_drawable_layer.hpp>
+#include <mbgl/math/wrap.hpp>
+
+#include <mbgl/style/layers/custom_drawable_layer.hpp>
+#include <mbgl/gfx/context.hpp>
+#include <mbgl/gfx/drawable.hpp>
+#include <mbgl/gfx/drawable_tweaker.hpp>
+#include <mbgl/gfx/shader_registry.hpp>
+#include <mbgl/gfx/color_mode.hpp>
+#include <mbgl/gfx/cull_face_mode.hpp>
+#include <mbgl/gfx/drawable_builder.hpp>
+
+#include <mbgl/shaders/shader_program_base.hpp>
+#include <mbgl/util/string_indexer.hpp>
+#include <mbgl/renderer/layer_tweaker.hpp>
+#include <mbgl/renderer/layer_group.hpp>
+#include <mbgl/renderer/render_tile.hpp>
+#include <mbgl/shaders/line_layer_ubo.hpp>
+#include <mbgl/util/string_indexer.hpp>
+#include <mbgl/util/convert.hpp>
+
+#if MLN_RENDER_BACKEND_METAL
+#include <mbgl/shaders/mtl/line.hpp>
+#endif // MLN_RENDER_BACKEND_METAL
+
+#include <memory>
+
+class MLNCustomDrawableLayerHost;
+
+@interface MLNCustomDrawableStyleLayer ()
+
+@property (nonatomic, readonly) mbgl::style::CustomDrawableLayer *rawLayer;
+
+@property (nonatomic, readonly, nullable) MLNMapView *mapView;
+
+@property (nonatomic, weak, readwrite) MLNStyle *style;
+
+@end
+
+@implementation MLNCustomDrawableStyleLayer
+
+- (instancetype)initWithIdentifier:(NSString *)identifier {
+    auto layer = std::make_unique<mbgl::style::CustomDrawableLayer>(identifier.UTF8String,
+                                                            std::make_unique<MLNCustomDrawableLayerHost>(self));
+    return self = [super initWithPendingLayer:std::move(layer)];
+}
+
+- (mbgl::style::CustomDrawableLayer *)rawLayer {
+    return (mbgl::style::CustomDrawableLayer *)super.rawLayer;
+}
+
+- (MLNMapView *)mapView {
+    if ([self.style.stylable isKindOfClass:[MLNMapView class]]) {
+        return (MLNMapView *)self.style.stylable;
+    }
+    return nil;
+}
+
+- (void)didMoveToMapView:(MLNMapView *)mapView {
+
+}
+
+- (void)willMoveFromMapView:(MLNMapView *)mapView {
+
+}
+
+- (void)setNeedsDisplay {
+    [self.mapView setNeedsRerender];
+}
+
+@end
+
+using namespace mbgl;
+using namespace mbgl::style;
+
+class TestDrawableTweaker : public gfx::DrawableTweaker {
```

Comment:
This appears to be duplicated in `test/api/custom_drawable_layer.test.cpp`. If this is intended to be used in a test, it should be relocated to `platform/darwin/test`.

Reply:
yes, this is similar to `test/api/custom_layer.test.cpp` versus `platform/darwin/src/MLNOpenGLStyleLayer.mm` + `platform/darwin/app/LimeGreenStyleLayer.m". 
As far as I can see, the former is meant to run as a platform independent test, while the latter has dependencies to the `darwin` platform and is meant to support the _demo_ in the iOS App. 
I'm refactoring to remove code duplication, but as I'm devising a friendlier API to be used in subsequent PRs and the whole _user code_ areas will change.

Reply:
I am relocating `platform/darwin/src/MLNCustomDrawableStyleLayer.mm` to `platform/darwin/app` as it is essentially example code, to be demonstrated by the `App`

Reply:
this looks to be much more work than I anticipated, a whole new objective C interface is needed. I'm cutting this PR here with the note that platforms are not yet supported.

Reply:
I'm removing the code from this branch and we'll expose the functionality at platform level when it's ready. 
the API test remains, since that uses the core directly through C++.